### PR TITLE
WIP: Use node id as a hint of which node to compare to

### PR DESCRIPTION
**Pull request by finnp**
Hey,

this pull request tries to implement an algorithm for #8. It should perform well on removal and addition of elements in a list. However for reordering it will be worse, especially if the list of dom nodes is inverted. But I guess there is not much we can do in that case.

The theoretical optimal dom modification would be similar to a "Shortest Edit Script" and quite expensive to perform. This algorithm here has multiple nested loops, but has a linear time.

I haven't yet figured out how "mixed content" should be treated, e.g.
```html
<ul>
    <li id="1">hi</li>
    <li>there</li>
    <li id="2"></li>
</ul>
````

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -49,32 +49,42 @@ function walk (newNode, oldNode) {
...
-  if (!newNode.childNodes || !oldNode.childNodes) return
+  var oldChildren = oldNode.childNodes
+  var newChildren = newNode.childNodes
+  if (!newChildren || !oldChildren) return
 
-  var newLength = newNode.childNodes.length
-  var oldLength = oldNode.childNodes.length
-  var length = Math.max(oldLength, newLength)
+  var newIndex = 0
+  for (var oldIndex = 0; oldIndex < oldChildren.length; oldIndex++) {
```
Comment (yoshuawuyts) [inline index.js]: Given that `oldChildren.length` is constant, we can evade doing an extra lookup on every loop:
```js
for (var oldIndex = 0, len = oldChildren.length; oldIndex < len; oldIndex++) {
```

File: index.js [hunk 2]
```diff
@@ -49,32 +49,42 @@ function walk (newNode, oldNode) {
...
-      if (oldChildNode) {
-        oldNode.removeChild(oldChildNode)
-        iOld--
-      }
-    } else if (!oldChildNode) {
-      if (retChildNode) {
-        oldNode.appendChild(retChildNode)
-        iNew--
+  function findNewChild () {
+    for (; newIndex < newChildren.length; newIndex++) {
```
Comment (yoshuawuyts) [inline index.js]: Here too, `newChildren` has a constant length

File: test.js [hunk 3]
```diff
@@ -252,6 +252,46 @@ tape('should skip over data-onload attributes at root', function (t) {
...
+
+    var c = nanomorph(a, b)
+    t.equal(oldFirst, c.children[0], 'first is equal')
+    t.equal(oldSecond, c.children[2], 'moved second is equal')
+    t.equal(oldThird, c.children[3], 'moved third is equal')
+    t.end()
+  })
+
+  t.test('remove an element', function (t) {
+    var a = html`<ul><li id="a"></li><li id="b"></li><li id="c"></li></ul>`
```
Comment (yoshuawuyts) [inline test.js]: Could you also multiline this example like the one above? Think it helps with readability. Thanks!

File: index.js (hunk 1 above)
Comment (yoshuawuyts) [inline index.js]: lol, nope not this one - apparently we modify the length of `oldChildren` - the next one is actually static tho

**Comment by yoshuawuyts** [comment]
Super neato patch! - really digging this :grin:

Did some super quick testing, and it doesn't appear to be making much of a difference in `choo/example` because even though the comparison might be cheaper - we still do the full comparison of all properties (which is the expensive part).

Even though this patch is sound, I wonder how much it'll improve performance. I think where it'll truly shine, is eliminating a class of errors for stateful elements (cached or otherwise) to prevent their placeholder return value (`el.isSameNode()`) from being rendered on accident.

So I think if we're gonna bench, we should bench with fancy, cached nodes so we actually get a perf benefit from comparing two similar nodes :sparkles:

File: test.js (hunk 3 above)
Comment (finnp) [inline test.js]: I need to debug this one. If I introduce newlines, there will be TextNodes in between the `li` and it will not work anymore.

Comment (yoshuawuyts) [inline test.js]: ah right, weirrd haha

File: index.js (hunk 1 above)
Comment (juliangruber) [inline index.js]: + `.length` is often optimized in engines too

File: test.js (hunk 3 above)
Comment (juliangruber) [inline test.js]: use `.childNodes` instead of `.children` then as those ignore text nodes

**Comment by juliangruber** [comment]
I'm giving this a try in a real world application now that hit exactly this bug, will report back here

**Comment by juliangruber** [comment]
works for me! 🎆

File: test.js (hunk 3 above)
Comment (yoshuawuyts) [inline test.js]: turns out `.children` ignores text nodes https://developer.mozilla.org/en-US/docs/Web/API/ParentNode/children

**Comment by finnp** [comment]
I was pretty sure that the TextNodes still caused bugs. Are you sure this was good to merge?

**Comment by juliangruber** [comment]
> I was pretty sure that the TextNodes still caused bugs. Are you sure this was good to merge?

as far as i can tell the bug was only in the test code, right?

**Comment by yoshuawuyts** [comment]
Think it might trigger anywhere, wrote a patch for Bel so random whitespace no longer creates text nodes; think this should help (:
- https://github.com/shama/bel/pull/68

**Comment by yoshuawuyts** [comment]
Ya, confirmed - that patch fixes our tests when made multiline (:

**Comment by finnp** [comment]
Awesome 👍

File: test.js (hunk 3 above)
Comment (juliangruber) [inline test.js]: oh, you're right `.children` is correct

**Comment by yoshuawuyts** [comment]
And now also https://github.com/shama/yo-yoify/pull/45 :sparkles:

**Pull request merged**